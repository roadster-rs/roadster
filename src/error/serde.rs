use crate::error::Error;

#[derive(Debug, Error)]
#[non_exhaustive]
pub enum SerdeError {
    #[error(transparent)]
    Json(#[from] serde_json::Error),

    #[error(transparent)]
    TomlDeserialize(#[from] toml::de::Error),

    #[error(transparent)]
    TomlSerialize(#[from] toml::ser::Error),

    #[error(transparent)]
    Other(#[from] Box<dyn std::error::Error + Send + Sync>),
}

impl From<serde_json::Error> for Error {
    fn from(value: serde_json::Error) -> Self {
        Self::Serde(SerdeError::from(value))
    }
}

impl From<toml::de::Error> for Error {
    fn from(value: toml::de::Error) -> Self {
        Self::Serde(SerdeError::from(value))
    }
}

impl From<toml::ser::Error> for Error {
    fn from(value: toml::ser::Error) -> Self {
        Self::Serde(SerdeError::from(value))
    }
}
