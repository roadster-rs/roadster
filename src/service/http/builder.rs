#[cfg(feature = "open-api")]
use crate::api::http::default_api_routes;
#[cfg(not(feature = "open-api"))]
use crate::api::http::default_routes;
use crate::app::context::AppContext;
use crate::app::App;
use crate::error::RoadsterResult;
use crate::service::http::initializer::default::default_initializers;
use crate::service::http::initializer::Initializer;
use crate::service::http::middleware::default::default_middleware;
use crate::service::http::middleware::Middleware;
use crate::service::http::service::{enabled, HttpService, NAME};
use crate::service::AppServiceBuilder;
#[cfg(feature = "open-api")]
use aide::axum::ApiRouter;
#[cfg(feature = "open-api")]
use aide::openapi::OpenApi;
#[cfg(feature = "open-api")]
use aide::transform::TransformOpenApi;
use anyhow::anyhow;
use async_trait::async_trait;
#[cfg(feature = "open-api")]
use axum::Extension;
use axum::Router;
use itertools::Itertools;
use std::collections::BTreeMap;
#[cfg(feature = "open-api")]
use std::sync::Arc;
use tracing::info;

pub struct HttpServiceBuilder<A: App + 'static> {
    context: AppContext<A::State>,
    router: Router<AppContext<A::State>>,
    #[cfg(feature = "open-api")]
    api_router: ApiRouter<AppContext<A::State>>,
    #[cfg(feature = "open-api")]
    api_docs: Box<dyn Fn(TransformOpenApi) -> TransformOpenApi + Send>,
    middleware: BTreeMap<String, Box<dyn Middleware<A::State>>>,
    initializers: BTreeMap<String, Box<dyn Initializer<A::State>>>,
}

impl<A: App> HttpServiceBuilder<A> {
    pub fn new(path_root: Option<&str>, context: &AppContext<A::State>) -> Self {
        // Normally, enabling a feature shouldn't remove things. In this case, however, we don't
        // want to include the default routes in the axum::Router if the `open-api` features is
        // enabled. Otherwise, we'll get a route conflict when the two routers are merged.
        #[cfg(not(feature = "open-api"))]
        let router = default_routes(path_root.unwrap_or_default(), context);
        #[cfg(feature = "open-api")]
        let router = Router::<AppContext<A::State>>::new();

        #[cfg(feature = "open-api")]
        let app_name = context.config().app.name.clone();
        Self {
            context: context.clone(),
            router,
            #[cfg(feature = "open-api")]
            api_router: default_api_routes(path_root.unwrap_or_default(), context),
            #[cfg(feature = "open-api")]
            api_docs: Box::new(move |api| {
                api.title(&app_name).description(&format!("# {}", app_name))
            }),
            middleware: default_middleware(context),
            initializers: default_initializers(context),
        }
    }

    #[cfg(test)]
    fn empty(context: &AppContext<A::State>) -> Self {
        Self {
            context: context.clone(),
            router: Router::<AppContext<A::State>>::new(),
            #[cfg(feature = "open-api")]
            api_router: ApiRouter::<AppContext<A::State>>::new(),
            #[cfg(feature = "open-api")]
            api_docs: Box::new(|op| op),
            middleware: Default::default(),
            initializers: Default::default(),
        }
    }

    pub fn router(mut self, router: Router<AppContext<A::State>>) -> Self {
        self.router = self.router.merge(router);
        self
    }

    #[cfg(feature = "open-api")]
    pub fn api_router(mut self, router: ApiRouter<AppContext<A::State>>) -> Self {
        self.router = self.router.merge(router);
        self
    }

    #[cfg(feature = "open-api")]
    pub fn api_docs(
        mut self,
        api_docs: Box<dyn Fn(TransformOpenApi) -> TransformOpenApi + Send>,
    ) -> Self {
        self.api_docs = api_docs;
        self
    }

    pub fn initializer<T>(mut self, initializer: T) -> RoadsterResult<Self>
    where
        T: Initializer<A::State> + 'static,
    {
        if !initializer.enabled(&self.context) {
            return Ok(self);
        }
        let name = initializer.name();
        if self
            .initializers
            .insert(name.clone(), Box::new(initializer))
            .is_some()
        {
            return Err(anyhow!("Initializer `{name}` was already registered").into());
        }
        Ok(self)
    }

    pub fn middleware<T>(mut self, middleware: T) -> RoadsterResult<Self>
    where
        T: Middleware<A::State> + 'static,
    {
        if !middleware.enabled(&self.context) {
            return Ok(self);
        }
        let name = middleware.name();
        if self
            .middleware
            .insert(name.clone(), Box::new(middleware))
            .is_some()
        {
            return Err(anyhow!("Middleware `{name}` was already registered").into());
        }
        Ok(self)
    }
}

#[async_trait]
impl<A: App> AppServiceBuilder<A, HttpService> for HttpServiceBuilder<A> {
    fn name(&self) -> String {
        NAME.to_string()
    }

    fn enabled(&self, app_context: &AppContext<A::State>) -> bool {
        enabled(app_context)
    }

    async fn build(self, context: &AppContext<A::State>) -> RoadsterResult<HttpService> {
        let router = self.router;

        #[cfg(feature = "open-api")]
        let (router, api) = {
            let mut api = OpenApi::default();
            let api_docs = self.api_docs;
            let api_router = self.api_router.finish_api_with(&mut api, api_docs);
            let router = router.merge(api_router);
            // Arc is very important here or we will face massive memory and performance issues
            let api = Arc::new(api);
            let router = router.layer(Extension(api.clone()));
            (router, api)
        };

        let router = router.with_state::<()>(context.clone());

        let initializers = self
            .initializers
            .values()
            .filter(|initializer| initializer.enabled(context))
            .sorted_by(|a, b| Ord::cmp(&a.priority(context), &b.priority(context)))
            .collect_vec();

        let router = initializers
            .iter()
            .try_fold(router, |router, initializer| {
                initializer.after_router(router, context)
            })?;

        let router = initializers
            .iter()
            .try_fold(router, |router, initializer| {
                initializer.before_middleware(router, context)
            })?;

        info!("Installing middleware. Note: the order of installation is the inverse of the order middleware will run when handling a request.");
        let router = self
            .middleware
            .values()
            .filter(|middleware| middleware.enabled(context))
            .sorted_by(|a, b| Ord::cmp(&a.priority(context), &b.priority(context)))
            // Reverse due to how Axum's `Router#layer` method adds middleware.
            .rev()
            .try_fold(router, |router, middleware| {
                info!(middleware=%middleware.name(), "Installing middleware");
                middleware.install(router, context)
            })?;

        let router = initializers
            .iter()
            .try_fold(router, |router, initializer| {
                initializer.after_middleware(router, context)
            })?;

        let router = initializers
            .iter()
            .try_fold(router, |router, initializer| {
                initializer.before_serve(router, context)
            })?;

        let service = HttpService {
            router,
            #[cfg(feature = "open-api")]
            api,
        };

        Ok(service)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::app::context::AppContext;
    use crate::app::MockApp;
    use crate::service::http::initializer::MockInitializer;
    use crate::service::http::middleware::MockMiddleware;

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn middleware() {
        // Arrange
        let context = AppContext::<()>::test(None, None, None).unwrap();
        let builder = HttpServiceBuilder::<MockApp>::empty(&context);

        let mut middleware = MockMiddleware::default();
        middleware.expect_enabled().returning(|_| true);
        middleware.expect_name().returning(|| "test".to_string());

        // Act
        let builder = builder.middleware(middleware).unwrap();

        // Assert
        assert_eq!(builder.middleware.len(), 1);
        assert!(builder.middleware.contains_key("test"));
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn middleware_not_enabled() {
        // Arrange
        let context = AppContext::<()>::test(None, None, None).unwrap();
        let builder = HttpServiceBuilder::<MockApp>::empty(&context);

        let mut middleware = MockMiddleware::default();
        middleware.expect_enabled().returning(|_| false);

        // Act
        let builder = builder.middleware(middleware).unwrap();

        // Assert
        assert!(builder.middleware.is_empty());
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    #[should_panic]
    fn middleware_already_registered() {
        // Arrange
        let context = AppContext::<()>::test(None, None, None).unwrap();
        let builder = HttpServiceBuilder::<MockApp>::empty(&context);

        let mut middleware = MockMiddleware::default();
        middleware.expect_name().returning(|| "test".to_string());
        let builder = builder.middleware(middleware).unwrap();

        let mut middleware = MockMiddleware::default();
        middleware.expect_name().returning(|| "test".to_string());

        // Act
        builder.middleware(middleware).unwrap();
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn initializer() {
        // Arrange
        let context = AppContext::<()>::test(None, None, None).unwrap();
        let builder = HttpServiceBuilder::<MockApp>::empty(&context);

        let mut initializer = MockInitializer::default();
        initializer.expect_enabled().returning(|_| true);
        initializer.expect_name().returning(|| "test".to_string());

        // Act
        let builder = builder.initializer(initializer).unwrap();

        // Assert
        assert_eq!(builder.initializers.len(), 1);
        assert!(builder.initializers.contains_key("test"));
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn initializer_not_enabled() {
        // Arrange
        let context = AppContext::<()>::test(None, None, None).unwrap();
        let builder = HttpServiceBuilder::<MockApp>::empty(&context);

        let mut initializer = MockInitializer::default();
        initializer.expect_enabled().returning(|_| false);

        // Act
        let builder = builder.initializer(initializer).unwrap();

        // Assert
        assert!(builder.initializers.is_empty());
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    #[should_panic]
    fn initializer_already_registered() {
        // Arrange
        let context = AppContext::<()>::test(None, None, None).unwrap();
        let builder = HttpServiceBuilder::<MockApp>::empty(&context);

        let mut initializer = MockInitializer::default();
        initializer.expect_name().returning(|| "test".to_string());
        let builder = builder.initializer(initializer).unwrap();

        let mut initializer = MockInitializer::default();
        initializer.expect_name().returning(|| "test".to_string());

        // Act
        builder.initializer(initializer).unwrap();
    }
}
