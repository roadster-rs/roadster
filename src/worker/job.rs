use crate::worker::backend::pg::processor::builder::PeriodicArgsJson;
use cron::Schedule;
use serde::{Deserialize, Serialize};
use serde_with::skip_serializing_none;
use std::hash::{DefaultHasher, Hash, Hasher};

// Todo: Not sure if this should be public yet.
#[derive(Serialize, Deserialize, bon::Builder)]
pub(crate) struct Job {
    pub(crate) metadata: JobMetadata,
    pub(crate) args: serde_json::Value,
}

// Todo: Not sure if this should be public yet.
#[skip_serializing_none]
#[derive(Serialize, Deserialize, bon::Builder)]
pub(crate) struct JobMetadata {
    pub(crate) worker_name: String,
    pub(crate) periodic: Option<PeriodicConfig>,
}

#[skip_serializing_none]
#[derive(Debug, Serialize, Deserialize, bon::Builder)]
pub(crate) struct PeriodicConfig {
    pub(crate) hash: u64,
    pub(crate) schedule: Schedule,
}

impl From<&PeriodicArgsJson> for Job {
    fn from(value: &PeriodicArgsJson) -> Self {
        let hash = periodic_hash(&value.worker_name, &value.schedule, &value.args);
        Job::builder()
            .args(value.args.clone())
            .metadata(
                JobMetadata::builder()
                    .worker_name(value.worker_name.clone())
                    .periodic(
                        PeriodicConfig::builder()
                            .hash(hash)
                            .schedule(value.schedule.clone())
                            .build(),
                    )
                    .build(),
            )
            .build()
    }
}

pub(crate) fn periodic_hash(
    worker_name: &str,
    schedule: &Schedule,
    value: &serde_json::Value,
) -> u64 {
    let mut hash = DefaultHasher::new();
    worker_name.hash(&mut hash);
    schedule.to_string().hash(&mut hash);
    value.hash(&mut hash);
    hash.finish()
}

#[cfg(test)]
mod tests {
    use crate::testing::snapshot::TestCase;
    use cron::Schedule;
    use insta::assert_snapshot;
    use rstest::{fixture, rstest};
    use std::str::FromStr;

    #[fixture]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case("a", Schedule::from_str("* * * * * *").unwrap(), serde_json::json!({"foo": "bar"}))]
    #[case("b", Schedule::from_str("*/10 * * * * *").unwrap(), serde_json::json!({"foo": "baz"}))]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn periodic_hash(
        _case: TestCase,
        #[case] name: &str,
        #[case] schedule: Schedule,
        #[case] value: serde_json::Value,
    ) {
        assert_snapshot!(super::periodic_hash(name, &schedule, &value));
    }
}
