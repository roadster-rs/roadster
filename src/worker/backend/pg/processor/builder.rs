use crate::app::context::AppContext;
use crate::error::RoadsterResult;
use crate::worker::Worker;
use crate::worker::backend::pg::processor::{
    PgProcessor, PgProcessorError, ProcessorInner, WorkerWrapper,
};
use axum_core::extract::FromRef;
use cron::Schedule;
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;
use tracing::{error, info};

#[non_exhaustive]
pub struct PgProcessorBuilder<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    pub(crate) inner: ProcessorInner<S>,
}

#[derive(bon::Builder)]
#[non_exhaustive]
pub struct PeriodicArgs<Args>
where
    Args: Send + Sync + Serialize + for<'de> Deserialize<'de>,
{
    args: Args,
    schedule: Schedule,
}

#[derive(Clone, bon::Builder, Eq, PartialEq)]
#[non_exhaustive]
pub(crate) struct PeriodicArgsJson {
    pub(crate) args: serde_json::Value,
    pub(crate) worker_name: String,
    pub(crate) schedule: Schedule,
}

impl Ord for PeriodicArgsJson {
    fn cmp(&self, other: &Self) -> Ordering {
        self.worker_name
            .cmp(&other.worker_name)
            .then(self.schedule.to_string().cmp(&other.schedule.to_string()))
            .then(
                serde_json::to_string(&self.args)
                    .unwrap_or_default()
                    .cmp(&serde_json::to_string(&other.args).unwrap_or_default()),
            )
    }
}

impl PartialOrd for PeriodicArgsJson {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl<S> PgProcessorBuilder<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    pub(crate) fn new(state: &S) -> Self {
        Self {
            inner: ProcessorInner {
                state: state.clone(),
                queues: Default::default(),
                workers: Default::default(),
                periodic_workers: Default::default(),
            },
        }
    }

    pub fn build(self) -> PgProcessor<S> {
        PgProcessor::new(self.inner)
    }

    pub fn register<W, Args, E>(mut self, worker: W) -> RoadsterResult<Self>
    where
        W: 'static + Worker<S, Args, Error = E>,
        Args: Send + Sync + Serialize + for<'de> Deserialize<'de>,
        // Todo: without this `'static`, we're getting an internal compiler error
        E: 'static + std::error::Error + Send + Sync,
    {
        let name = W::name();
        info!(name, "Registering PG worker");

        self.register_internal(worker, name, false)?;

        Ok(self)
    }

    pub fn register_periodic<W, Args, E>(
        mut self,
        worker: W,
        periodic_args: PeriodicArgs<Args>,
    ) -> RoadsterResult<Self>
    where
        W: 'static + Worker<S, Args, Error = E>,
        Args: Send + Sync + Serialize + for<'de> Deserialize<'de>,
        // Todo: without this `'static`, we're getting an internal compiler error
        E: 'static + std::error::Error + Send + Sync,
    {
        let name = W::name();
        info!(name, "Registering periodic PG worker");

        self.register_internal(worker, name.clone(), true)?;

        let periodic_args = PeriodicArgsJson::builder()
            .args(serde_json::to_value(periodic_args.args)?)
            .worker_name(name.clone())
            .schedule(periodic_args.schedule)
            .build();

        if let Some(replaced) = self.inner.periodic_workers.replace(periodic_args) {
            return Err(PgProcessorError::AlreadyRegisteredPeriodic(
                replaced.worker_name,
                replaced.schedule,
                replaced.args,
            )
            .into());
        }

        Ok(self)
    }

    fn register_internal<W, Args, E>(
        &mut self,
        worker: W,
        name: String,
        skip_duplicate: bool,
    ) -> RoadsterResult<()>
    where
        W: 'static + Worker<S, Args, Error = E>,
        Args: Send + Sync + Serialize + for<'de> Deserialize<'de>,
        // Todo: without this `'static`, we're getting an internal compiler error
        E: 'static + std::error::Error + Send + Sync,
    {
        let context = AppContext::from_ref(&self.inner.state);
        let enqueue_config = &context.config().service.worker.enqueue_config;
        let worker_enqueue_config = W::enqueue_config(&self.inner.state);

        let queue = worker_enqueue_config
            .queue
            .as_ref()
            .or(enqueue_config.queue.as_ref());
        let queue = if let Some(queue) = queue {
            queue
        } else {
            error!(
                worker_name = W::name(),
                "Unable to register worker, no queue configured"
            );
            return Err(PgProcessorError::NoQueue(W::name()).into());
        };
        self.inner.queues.insert(queue.to_owned());

        if self
            .inner
            .workers
            .insert(
                name.clone(),
                WorkerWrapper::new(&self.inner.state, worker, worker_enqueue_config)?,
            )
            .is_some()
            && !skip_duplicate
        {
            return Err(PgProcessorError::AlreadyRegistered(name).into());
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    mod periodic_args {
        use crate::worker::backend::pg::processor::builder::PeriodicArgsJson;
        use crate::worker::job::Job;
        use cron::Schedule;
        use insta::assert_json_snapshot;
        use rstest::{fixture, rstest};
        use std::str::FromStr;

        #[fixture]
        #[cfg_attr(coverage_nightly, coverage(off))]
        fn periodic_args_json() -> PeriodicArgsJson {
            PeriodicArgsJson::builder()
                .worker_name("a".to_string())
                .schedule(Schedule::from_str("* * * * * *").unwrap())
                .args(serde_json::json!({"foo": "bar"}))
                .build()
        }

        #[rstest]
        #[cfg_attr(coverage_nightly, coverage(off))]
        fn periodic_args_json_ord_name(periodic_args_json: PeriodicArgsJson) {
            let mut b = periodic_args_json.clone();
            b.worker_name = "b".to_string();
            assert!(periodic_args_json < b);
        }

        #[rstest]
        #[cfg_attr(coverage_nightly, coverage(off))]
        fn periodic_args_json_ord_schedule(periodic_args_json: PeriodicArgsJson) {
            let mut b = periodic_args_json.clone();
            b.schedule = Schedule::from_str("*/10 * * * * *").unwrap();
            assert!(periodic_args_json < b);
        }

        #[rstest]
        #[cfg_attr(coverage_nightly, coverage(off))]
        fn periodic_args_json_ord_args(periodic_args_json: PeriodicArgsJson) {
            let mut b = periodic_args_json.clone();
            b.args = serde_json::json!({"foo": "baz"});
            assert!(periodic_args_json < b);
        }

        #[rstest]
        #[cfg_attr(coverage_nightly, coverage(off))]
        fn job_from_periodic_args(periodic_args_json: PeriodicArgsJson) {
            let job = Job::from(&periodic_args_json);
            assert_json_snapshot!(job);
        }

        #[rstest]
        #[cfg_attr(coverage_nightly, coverage(off))]
        fn job_from_periodic_args_hash(periodic_args_json: PeriodicArgsJson) {
            let job = Job::from(&periodic_args_json);
            let hash = crate::worker::job::periodic_hash(
                &job.metadata.worker_name,
                &job.metadata.periodic.as_ref().unwrap().schedule,
                &job.args,
            );
            assert_eq!(hash, job.metadata.periodic.unwrap().hash);
        }
    }
}
