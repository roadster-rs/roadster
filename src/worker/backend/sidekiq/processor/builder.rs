use crate::app::context::AppContext;
use crate::error::RoadsterResult;
use crate::worker::backend::sidekiq::processor::{
    SidekiqProcessor, SidekiqProcessorError, SidekiqProcessorInner,
};
use crate::worker::backend::sidekiq::roadster_worker::RoadsterWorker;
use crate::worker::{PeriodicArgs, PeriodicArgsJson, RegisterSidekiqFn, Worker, WorkerWrapper};
use axum_core::extract::FromRef;
use serde::{Deserialize, Serialize};
use sidekiq::Processor;
use std::collections::{BTreeMap, BTreeSet};
use tracing::{error, info};

#[non_exhaustive]
pub struct SidekiqProcessorBuilder<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    pub(crate) state: S,
    // Todo: we may need to register directly on the processor instead of waiting to register
    //  until later, depending on if `RoadsterWorker` needs the `W` type param.
    // todo: store a closure to register the worker in order to keep the type?
    pub(crate) processor: Option<::sidekiq::Processor>,
    pub(crate) queues: BTreeSet<String>,
    pub(crate) workers: BTreeMap<String, (WorkerWrapper<S>, RegisterSidekiqFn<S>)>,
    pub(crate) periodic_workers: BTreeSet<PeriodicArgsJson>,
}

impl<S> SidekiqProcessorBuilder<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    pub(crate) fn new(state: &S) -> Self {
        Self {
            state: state.clone(),
            queues: Default::default(),
            processor: Default::default(),
            workers: Default::default(),
            periodic_workers: Default::default(),
        }
    }

    pub fn build(mut self) -> SidekiqProcessor<S> {
        // todo: create processor if it wasn't provided.
        if let Some(processor) = self.processor.as_mut() {
            for (worker, register_fn) in self.workers.into_values() {
                register_fn(&self.state, processor, worker);
            }
        }

        SidekiqProcessor::new(SidekiqProcessorInner {
            state: self.state,
            processor: self.processor,
            queues: self.queues,
            periodic_workers: self.periodic_workers,
        })
    }

    pub fn with_processor(mut self, processor: ::sidekiq::Processor) -> Self {
        self.processor = Some(processor);
        self
    }

    pub fn register<W, Args, E>(mut self, worker: W) -> RoadsterResult<Self>
    where
        W: 'static + Worker<S, Args, Error = E>,
        Args: 'static + Send + Sync + Serialize + for<'de> Deserialize<'de>,
        E: 'static + std::error::Error + Send + Sync,
    {
        let name = W::name();
        info!(name, "Registering Sidekiq worker");

        self.register_internal(worker, name, false)?;

        Ok(self)
    }

    pub fn register_periodic<W, Args, E>(
        mut self,
        worker: W,
        periodic_args: PeriodicArgs<Args>,
    ) -> RoadsterResult<Self>
    where
        W: 'static + Worker<S, Args, Error = E>,
        Args: 'static + Send + Sync + Serialize + for<'de> Deserialize<'de>,
        E: 'static + std::error::Error + Send + Sync,
    {
        let name = W::name();
        info!(name, "Registering periodic PG worker");

        self.register_internal(worker, name.clone(), true)?;

        let periodic_args = PeriodicArgsJson::builder()
            .args(serde_json::to_value(periodic_args.args)?)
            .worker_name(name.clone())
            .schedule(periodic_args.schedule)
            .build();

        if let Some(replaced) = self.periodic_workers.replace(periodic_args) {
            return Err(SidekiqProcessorError::AlreadyRegisteredPeriodic(
                replaced.worker_name,
                replaced.schedule,
                replaced.args,
            )
            .into());
        }

        Ok(self)
    }

    fn register_internal<W, Args, E>(
        &mut self,
        worker: W,
        name: String,
        skip_duplicate: bool,
    ) -> RoadsterResult<()>
    where
        W: 'static + Worker<S, Args, Error = E>,
        Args: 'static + Send + Sync + Serialize + for<'de> Deserialize<'de>,
        E: 'static + std::error::Error + Send + Sync,
    {
        let context = AppContext::from_ref(&self.state);
        let enqueue_config = &context.config().service.worker.enqueue_config;
        let worker_enqueue_config = W::enqueue_config(&self.state);

        let queue = worker_enqueue_config
            .queue
            .as_ref()
            .or(enqueue_config.queue.as_ref());
        let queue = if let Some(queue) = queue {
            queue
        } else {
            error!(
                worker_name = W::name(),
                "Unable to register worker, no queue configured"
            );
            return Err(SidekiqProcessorError::NoQueue(W::name()).into());
        };
        self.queues.insert(queue.to_owned());

        // Todo: impl something similar for periodic jobs?
        let register_sidekiq = Box::new(
            move |state: &S, processor: &mut Processor, worker_wrapper: WorkerWrapper<S>| {
                let roadster_worker = RoadsterWorker::<S, W, Args, E>::new(state, worker_wrapper);
                processor.register(roadster_worker);
            },
        );

        if self
            .workers
            .insert(
                name.clone(),
                (
                    WorkerWrapper::new(&self.state, worker, worker_enqueue_config)?,
                    register_sidekiq,
                ),
            )
            .is_some()
            && !skip_duplicate
        {
            return Err(SidekiqProcessorError::AlreadyRegistered(name).into());
        }

        Ok(())
    }
}
