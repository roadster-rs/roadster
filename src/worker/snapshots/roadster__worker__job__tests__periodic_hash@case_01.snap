---
source: src/worker/job.rs
expression: "super::periodic_hash(name, &schedule, &value)"
---
11189630487790136136
